Support for Publicly Shareable Collections

ikreymer opened the pull request:
This PR adds a 'isPublic' toggle to collections. Public collections are accessible via a new 
`/api/orgs/<orgid>/collections/<coll_id>/public/replay.json` endpoint which is accessible without login and has CORS for any origin.

The PR borrows from previous work in #926 and #994. Closes #985, focusing on just public collections.

UI Changes:

Private collection options:
<img width="1552" alt="Screen Shot 2023-08-02 at 9 46 30 AM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/f9f9d4cb-7b15-4470-92a0-bc604fa6f7b1">

Public collection options: 
'Go to Public View' opens in ReplayWeb.page
<img width="1552" alt="Screen Shot 2023-08-02 at 9 46 45 AM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/2a903778-63bf-4d9a-92f5-19c8cebcd10b">

View Embed (from #994):
<img width="1552" alt="Screen Shot 2023-08-02 at 9 49 28 AM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/b918d863-4da5-4dc6-9bd8-b68f00e211b8">

Also 'Publicly Accessible' switch on metadata screen:
<img width="1552" alt="Screen Shot 2023-08-02 at 9 51 12 AM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/14074160-c318-49ab-83c9-15f0920dec87">

Collections view with icons (final):
<img width="1207" alt="Screen Shot 2023-08-03 at 7 10 03 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/600a86c7-bcc4-4169-bf9e-f93878b52d60">


Still needed (on RWP):
- Ensure standalone RWP automatically updates when new crawls are added
- Additional RWP improvements for (thumbnails, etc..)

ikreymer committed:
collections: support toggling collections public/private, viewable via RWP
- backend: add 'public' to collection model, support patching to update
- backend: add .../collections/<id>/public/replay.json for public access
- backend: add CORS handling for public endpoint
- frontend: support 'make public / make private' dropdown actions
- frontend: link to replayweb.page for standalone browsing
- frontend: add embed code popup

ikreymer committed:
tweak text

Shrinks99 commented on frontend/src/pages/org/collection-detail.ts:
```diff
@@ -86,9 +89,33 @@ export class CollectionDetail extends LiteElement {
+              View Embed Code
+            </sl-button>
+          `
+        )}
```
Sorry for the awful diff.  This moves the globe out of the H1, adds a `<div>` to keep them both positioned correctly, replaces the title attribute with a shoelace tooltip, and removes the padding on the embed button to reduce layout shift.

EDIT: Actually I'm going to hide this comment, this is a big enough block of code that should be linted / pre-commit formatted and submitted as a real commit.

```suggestion
      <div class="flex items-center gap-2 w-full mb-2 md:mb-0">
      ${when(
        this.collection?.isPublic,
        () => html`
          <sl-tooltip content=${msg("Publicly Accessible")}>
            <sl-icon
              style="font-size: 16px; color: var(--sl-color-success-600)"
              name="globe2"
              label="${msg("Publicly Accessible")}"
            ></sl-icon>
          </sl-tooltip>
        `
      )}
        <h1
          class="flex-1 min-w-0 text-xl font-semibold leading-7 truncate"
        >
          ${this.collection?.name ||
          html`<sl-skeleton class="w-96"></sl-skeleton>`}
        </h1>
        </div>
        ${when(
          this.collection?.isPublic,
          () => html`
            <sl-button
              size="small"
              @click=${() => (this.showEmbedInfo = true)}
            >
              <sl-icon name="code-slash"></sl-icon>
              View Embed Code
            </sl-button>
          `
        )}
```

Shrinks99 replied:
Don't want to deal with this via a suggestion, bad diff, should be pre-commit formatted.  Let me know what workflow is preferable here.

tw4l reviewed (approved):
Tested locally and working great

Shrinks99 commented on frontend/src/pages/org/collection-detail.ts:
```diff
@@ -190,6 +293,35 @@ export class CollectionDetail extends LiteElement {
+              `
+            : html`
+                <sl-menu-item style="--sl-color-neutral-700: var(--success)">
+                  <sl-icon name="box-arrow-up-left" slot="prefix"></sl-icon>
```
Could we change to the right arrow variant of this and also swap it out in the other areas of the app that use menu items as links (technically not ideal) later?  I like the box arrows better for going elsewhere but left signifies going back.

<img width="249" alt="Screenshot 2023-08-02 at 5 18 33 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5672810/a4dea5f6-1de2-40d0-9ae2-55209171e6ff">

```suggestion
                  <sl-icon name="box-arrow-up-right" slot="prefix"></sl-icon>
```

ikreymer replied:
Done in separate commit

SuaYoo reviewed (approved):
Nice!

Shrinks99 committed:
Replace `title` with `sl-tooltip`

- Adds tooltip to globe icon
- Moves globe icon out of the header and places both in a div
- Removes padding that caused layout shift issues

ikreymer committed:
Merge branch 'main' into public-collections

ikreymer committed:
frontend: tweak icons for public/private, label as 'make shareable', add shareable action items to collection list view

ikreymer committed:
eye-slash -> eye-slash-fill

Shrinks99 committed:
Change the menu option to use eye-slash (unfilled)

Line icons used for actions throughout the app

Shrinks99 committed:
Revert margin change edited elsewhere in branch

- Fixes merge of main re-adding the margin change I removed

Shrinks99 committed:
Switches eye-slash (unfilled) for list controls

ikreymer committed:
readd warn color on menu items, add tooltips for list view

ikreymer committed:
Merge branch 'public-collections' of github.com:webrecorder/browsertrix-cloud into public-collections

Shrinks99 reviewed (approved):
Looks great!
